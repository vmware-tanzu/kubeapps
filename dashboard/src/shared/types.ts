// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { JSONSchemaType } from "ajv";
import { RouterState } from "connected-react-router";
import {
  AvailablePackageDetail,
  AvailablePackageSummary,
  GetAvailablePackageSummariesResponse,
  InstalledPackageDetail,
  InstalledPackageSummary,
  PackageAppVersion,
  ResourceRef,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import {
  DockerCredentials,
  OpaqueCredentials,
  PackageRepositoryAuth_PackageRepositoryAuthType,
  SshCredentials,
  TlsCertKey,
  UsernamePassword,
} from "gen/kubeappsapis/core/packages/v1alpha1/repositories";
import { Plugin } from "gen/kubeappsapis/core/plugins/v1alpha1/plugins";
import { HelmPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/helm/packages/v1alpha1/helm";
import { KappControllerPackageRepositoryCustomDetail } from "gen/kubeappsapis/plugins/kapp_controller/packages/v1alpha1/kapp_controller";
import { IOperatorsState } from "reducers/operators";
import { Subscription } from "rxjs";
import { IAuthState } from "../reducers/auth";
import { IClustersState } from "../reducers/cluster";
import { IConfigState } from "../reducers/config";
import { IPackageRepositoryState } from "../reducers/repos";
import { RpcError } from "./RpcError";

export class CustomError extends Error {
  public causes: Error[] | undefined;
  // The constructor is defined so we can later on compare the returned object
  // via err.constructor  == FOO
  constructor(message?: string, causes?: Error[]) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
    this.causes = causes;
    this.checkCauses();
  }
  // Workaround used until RPC code (unary) throws a custom rpc error
  // Check if any RPC error is among the causes
  private checkCauses() {
    if (!this.causes) return;
    for (let i = 0; i < this.causes.length; i++) {
      const cause = this.causes[i];
      if (RpcError.isRpcError(cause)) {
        this.causes[i] = new RpcError(cause);
      }
    }
  }
}

// For 4XX HTTP-alike errors
export class ClientNetworkError extends CustomError {}
// 400
export class BadRequestNetworkError extends ClientNetworkError {}
// 401
export class UnauthorizedNetworkError extends ClientNetworkError {}
// 403
export class ForbiddenNetworkError extends ClientNetworkError {}
// 404
export class NotFoundNetworkError extends ClientNetworkError {}
// 408
export class RequestTimeoutNetworkError extends ClientNetworkError {}
// 409
export class ConflictNetworkError extends ClientNetworkError {}
// 422
export class UnprocessableEntityError extends ClientNetworkError {}
// 429
export class TooManyRequestsNetworkError extends ClientNetworkError {}

// For 5XX HTTP-alike errors
export class ServerNetworkError extends CustomError {}
// 500
export class InternalServerNetworkError extends ServerNetworkError {}
// 501
export class NotImplementedNetworkError extends ServerNetworkError {}
// 503
export class ServerUnavailableNetworkError extends ServerNetworkError {}
// 504
export class GatewayTimeoutNetworkError extends ServerNetworkError {}

// Application-level errors
export class FetchError extends CustomError {}

export class FetchWarning extends CustomError {}

export class CreateError extends CustomError {}

export class UpgradeError extends CustomError {}

export class RollbackError extends CustomError {}

export class DeleteError extends CustomError {}

export type DeploymentEvent = "install" | "upgrade";

export interface IReceivePackagesActionPayload {
  response: GetAvailablePackageSummariesResponse;
  paginationToken: string;
}

export interface IPackageState {
  isFetching: boolean;
  hasFinishedFetching: boolean;
  nextPageToken: string;
  selected: {
    error?: FetchError | Error;
    availablePackageDetail?: AvailablePackageDetail;
    pkgVersion?: string;
    appVersion?: string;
    versions: PackageAppVersion[];
    readme?: string;
    readmeError?: string;
    values?: string;
    schema?: JSONSchemaType<any>;
  };
  items: AvailablePackageSummary[];
  categories: string[];
  size: number;
}

export interface IServiceSpec {
  ports: IPort[];
  clusterIP: string;
  type: string;
}

export interface IServiceStatus {
  loadBalancer: {
    ingress?: Array<{ ip?: string; hostname?: string }>;
  };
}

export interface IPort {
  name: string;
  port: number;
  protocol: string;
  targetPort: string;
  nodePort: string;
}

export interface IHTTPIngressPath {
  path: string;
}

export interface IIngressHTTP {
  paths: IHTTPIngressPath[];
}

export interface IIngressRule {
  host: string;
  http: IIngressHTTP;
}

export interface IIngressTLS {
  hosts: string[];
}

export interface IIngressSpec {
  rules: IIngressRule[];
  tls?: IIngressTLS[];
  backend?: {
    serviceName: string;
    servicePort: number;
  };
}

export interface IResourceMetadata {
  name: string;
  namespace: string;
  annotations: { [key: string]: string };
  ownerReferences?: Array<{
    apiVersion: string;
    blockOwnerDeletion: string;
    kind: string;
    name: string;
    uid: string;
  }>;
  creationTimestamp: string;
  selfLink: string;
  resourceVersion: string;
  deletionTimestamp?: string;
  uid: string;
}

export interface IResource {
  apiVersion: string;
  kind: string;
  type: string;
  spec: any;
  status: any;
  metadata: IResourceMetadata;
}

export interface ISecret {
  apiVersion: string;
  kind: string;
  type: string;
  data: { [s: string]: string };
  metadata: IResourceMetadata;
}

export interface IPackageManifestChannel {
  name: string;
  currentCSV: string;
  currentCSVDesc: {
    annotations: {
      "alm-examples": string;
      capabilities: string;
      categories: string;
      certified: string;
      containerImage: string;
      createdAt: string;
      description: string;
      repository: string;
      support: string;
    };
    description: string;
    displayName: string;
    provider: {
      name: string;
    };
    version: string;
    installModes: [
      { supported: boolean; type: "OwnNamespace" },
      { supported: boolean; type: "SingleNamespace" },
      { supported: boolean; type: "MultiNamespace" },
      { supported: boolean; type: "AllNamespaces" },
    ];
    customresourcedefinitions: {
      owned: Array<{
        description: string;
        displayName: string;
        kind: string;
        name: string;
        version: string;
      }>;
    };
  };
}

export interface IPackageManifestStatus {
  catalogSource: string;
  catalogSourceDisplayName: string;
  catalogSourceNamespace: string;
  catalogSourcePublisher: string;
  provider: {
    name: string;
  };
  defaultChannel: string;
  channels: IPackageManifestChannel[];
}

export interface IPackageManifest extends IResource {
  status: IPackageManifestStatus;
}

export interface IClusterServiceVersionCRDResource {
  kind: string;
  name: string;
  version: string;
}

export interface IClusterServiceVersionCRD {
  description: string;
  displayName: string;
  kind: string;
  name: string;
  version: string;
  resources: IClusterServiceVersionCRDResource[];
  specDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
  statusDescriptors: Array<{
    description: string;
    displayName: string;
    path: string;
    "x-descriptors": string[];
  }>;
}

export interface IClusterServiceVersionSpec {
  apiservicedefinitions: any;
  customresourcedefinitions: {
    owned?: IClusterServiceVersionCRD[];
  };
  description: string;
  displayName: string;
  icon: Array<{
    base64data: string;
    mediatype: string;
  }>;
  install: any;
  installModes: [
    { supported: boolean; type: "OwnNamespace" },
    { supported: boolean; type: "SingleNamespace" },
    { supported: boolean; type: "MultiNamespace" },
    { supported: boolean; type: "AllNamespaces" },
  ];
  keywords: string[];
  labels: any;
  links: Array<{
    name: string;
    url: string;
  }>;
  maintainers: Array<{
    email: string;
    name: string;
  }>;
  maturity: string;
  provider: {
    name: string;
  };
  selector: any;
  version: string;
}

export interface IClusterServiceVersion extends IResource {
  spec: IClusterServiceVersionSpec;
}

export interface IPkgRepositoryFilter {
  jq: string;
  variables?: { [key: string]: string };
}

export interface IInstalledPackageState {
  isFetching: boolean;
  error?: FetchError | CreateError | UpgradeError | RollbackError | DeleteError;
  items: InstalledPackageDetail[];
  listOverview?: InstalledPackageSummary[];
  selected?: CustomInstalledPackageDetail;
  // TODO(agamez): add tests for this new state field
  selectedDetails?: AvailablePackageDetail;
  resourceRefs?: ResourceRef[];
}

export interface IStoreState {
  router: RouterState;
  apps: IInstalledPackageState;
  auth: IAuthState;
  packages: IPackageState;
  config: IConfigState;
  kube: IKubeState;
  repos: IPackageRepositoryState;
  clusters: IClustersState;
  operators: IOperatorsState;
}

interface IK8sResource {
  apiVersion: string;
  kind: string;
}

/** @see https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-and-simple-kinds */
export interface IK8sList<I, M> extends IK8sResource {
  items: I[];
  metadata?: {
    resourceVersion?: string;
    selfLink?: string; // Not in docs, but seems to exist everywhere
  } & M;
}

export interface IRBACRole {
  apiGroup: string;
  namespace?: string;
  clusterWide?: boolean;
  resource: string;
  verbs: string[];
}

export interface IKubeItem<T> {
  isFetching: boolean;
  item?: T;
  error?: Error;
}

export interface IKind {
  apiVersion: string;
  plural: string;
  namespaced: boolean;
}

export interface IKubeState {
  items: { [s: string]: IKubeItem<IResource | IK8sList<IResource, {}>> };
  subscriptions: { [s: string]: Subscription };
  // TODO(minelson): Remove kinds and kindsError once the operator support is
  // removed from the dashboard or replaced with a plugin.
  kinds: { [kind: string]: IKind };
  kindsError?: Error;
}

export interface IBasicFormParam {
  path: string;
  type?: "string" | "number" | "integer" | "boolean" | "object" | "array" | "null" | "any";
  value?: any;
  title?: string;
  minimum?: number;
  maximum?: number;
  render?: string;
  description?: string;
  customComponent?: object;
  enum?: string[];
  hidden?:
    | {
        event: DeploymentEvent;
        path: string;
        value: string;
        conditions: Array<{
          event: DeploymentEvent;
          path: string;
          value: string;
        }>;
        operator: string;
      }
    | string;
  children?: IBasicFormParam[];
}
export interface IBasicFormSliderParam extends IBasicFormParam {
  sliderMin?: number;
  sliderMax?: number;
  sliderStep?: number;
  sliderUnit?: string;
}

export interface CustomInstalledPackageDetail extends InstalledPackageDetail {
  revision: number;
}

//  enum for the type of package repository storage
export enum RepositoryStorageTypes {
  PACKAGE_REPOSITORY_STORAGE_HELM = "helm",
  PACKAGE_REPOSITORY_STORAGE_OCI = "oci",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_INLINE = "inline",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMAGE = "image",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_IMGPKGBUNDLE = "imgpkgBundle",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_HTTP = "http",
  PACKAGE_REPOSITORY_STORAGE_CARVEL_GIT = "git",
}

export enum PluginNames {
  PACKAGES_HELM = "helm.packages",
  PACKAGES_FLUX = "fluxv2.packages",
  PACKAGES_KAPP = "kapp_controller.packages",
}

export interface IPkgRepoFormData {
  authMethod: PackageRepositoryAuth_PackageRepositoryAuthType;
  // kubeapps-managed secrets
  authHeader: string; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BEARER || type == PACKAGE_REPOSITORY_AUTH_TYPE_AUTHORIZATION_HEADER
  basicAuth: UsernamePassword; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_BASIC_AUTH
  dockerRegCreds: DockerCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_DOCKER_CONFIG_JSON
  sshCreds: SshCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_SSH
  opaqueCreds: OpaqueCredentials; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_OPAQUE
  tlsCertKey: TlsCertKey; // used if type == PACKAGE_REPOSITORY_AUTH_TYPE_TLS
  // user-managed secrets
  secretAuthName: string;
  secretTLSName: string;
  // rest of the parameters
  customCA: string;
  description: string;
  interval: string;
  name: string;
  passCredentials: boolean;
  plugin: Plugin;
  skipTLS: boolean;
  type: string;
  url: string;
  // add more types if necessary
  customDetail?: Partial<
    HelmPackageRepositoryCustomDetail | KappControllerPackageRepositoryCustomDetail
  >;
  namespace: string;
  isNamespaceScoped: boolean;
}
