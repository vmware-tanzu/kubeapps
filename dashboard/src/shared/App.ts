import {
  AvailablePackageDetail,
  AvailablePackageReference,
  Context,
  CreateInstalledPackageRequest,
  InstalledPackageReference,
  ReconciliationOptions,
  VersionReference,
} from "gen/kubeappsapis/core/packages/v1alpha1/packages";
import * as url from "shared/url";
import { axiosWithAuth } from "./AxiosInstance";
import { KubeappsGrpcClient } from "./KubeappsGrpcClient";

export const KUBEOPS_ROOT_URL = "api/kubeops/v1";
export class App {
  private static client = () => new KubeappsGrpcClient().getPackagesServiceClientImpl();

  public static async GetInstalledPackageSummaries(
    cluster: string,
    namespace?: string,
    page?: number,
    size?: number,
  ) {
    return await this.client().GetInstalledPackageSummaries({
      context: { cluster: cluster, namespace: namespace },
      paginationOptions: { pageSize: size || 0, pageToken: page?.toString() || "0" },
    });
  }

  public static async GetInstalledPackageDetail(installedPackageRef?: InstalledPackageReference) {
    return await this.client().GetInstalledPackageDetail({
      installedPackageRef: installedPackageRef,
    });
  }

  public static async createInstalledPackage(
    targetContext: Context,
    name: string,
    availablePackageRef: AvailablePackageReference,
    pkgVersionReference: VersionReference,
    values?: string,
    reconciliationOptions?: ReconciliationOptions,
  ) {
    return await this.client().CreateInstalledPackage({
      name,
      values,
      targetContext,
      availablePackageRef,
      pkgVersionReference,
      reconciliationOptions,
    } as CreateInstalledPackageRequest);
  }

  public static async upgrade(
    installedPackageRef: InstalledPackageReference,
    chartNamespace: string,
    availablePackageDetail: AvailablePackageDetail,
    values?: string,
  ) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(endpoint, {
      appRepositoryResourceName:
        availablePackageDetail.availablePackageRef?.identifier.split("/")[0],
      appRepositoryResourceNamespace: chartNamespace,
      chartName: decodeURIComponent(availablePackageDetail.name),
      releaseName: installedPackageRef.identifier,
      values,
      version: availablePackageDetail.version?.pkgVersion,
    });
    return data;
  }

  public static async rollback(installedPackageRef: InstalledPackageReference, revision: number) {
    const endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    const { data } = await axiosWithAuth.put(
      endpoint,
      {},
      {
        params: {
          action: "rollback",
          revision,
        },
      },
    );
    return data;
  }

  public static async delete(installedPackageRef: InstalledPackageReference, purge: boolean) {
    let endpoint = url.kubeops.releases.get(
      installedPackageRef.context?.cluster ?? "",
      installedPackageRef.context?.namespace ?? "",
      installedPackageRef.identifier,
    );
    if (purge) {
      endpoint += "?purge=true";
    }
    const { data } = await axiosWithAuth.delete(endpoint);
    return data;
  }

  // TODO(agamez): remove it once we return the generated resources as part of the InstalledPackageDetail.
  public static async getRelease(installedPackageRef?: InstalledPackageReference) {
    const { data } = await axiosWithAuth.get<{ data: { manifest: any } }>(
      url.kubeops.releases.get(
        installedPackageRef?.context?.cluster ?? "",
        installedPackageRef?.context?.namespace ?? "",
        installedPackageRef?.identifier ?? "",
      ),
    );
    return data.data;
  }
}
