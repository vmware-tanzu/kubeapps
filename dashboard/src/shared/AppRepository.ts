import { axiosWithAuth } from "./AxiosInstance";
import { IAppRepositoryFilter, ICreateAppRepositoryResponse } from "./types";
import * as url from "./url";

export class AppRepository {
  public static async list(cluster: string, namespace: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.list(cluster, namespace));
    return appRepository;
  }

  public static async get(cluster: string, namespace: string, name: string) {
    const {
      data: { appRepository },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return appRepository;
  }

  public static async getSecretForRepo(cluster: string, namespace: string, name: string) {
    const {
      data: { secret },
    } = await axiosWithAuth.get<any>(url.backend.apprepositories.get(cluster, namespace, name));
    return secret;
  }

  public static async resync(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.post(
      url.backend.apprepositories.refresh(cluster, namespace, name),
      null,
    );
    return data;
  }

  public static async update(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.put<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.update(cluster, namespace, name),
      {
        appRepository: {
          name,
          repoURL,
          type,
          description,
          authHeader,
          authRegCreds,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async delete(cluster: string, namespace: string, name: string) {
    const { data } = await axiosWithAuth.delete(
      url.backend.apprepositories.delete(cluster, namespace, name),
    );
    return data;
  }

  // create uses the kubeapps backend API
  // TODO(mnelson) Update other endpoints to similarly use the backend API, removing the need
  // for direct k8s api access (for this resource, at least).
  public static async create(
    cluster: string,
    name: string,
    namespace: string,
    repoURL: string,
    type: string,
    description: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    syncJobPodTemplate: any,
    registrySecrets: string[],
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
    filter?: IAppRepositoryFilter,
  ) {
    const { data } = await axiosWithAuth.post<ICreateAppRepositoryResponse>(
      url.backend.apprepositories.create(cluster, namespace),
      {
        appRepository: {
          name,
          repoURL,
          authHeader,
          authRegCreds,
          type,
          description,
          customCA,
          syncJobPodTemplate,
          registrySecrets,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
          filterRule: filter,
        },
      },
    );
    return data;
  }

  public static async validate(
    cluster: string,
    namespace: string,
    repoURL: string,
    type: string,
    authHeader: string,
    authRegCreds: string,
    customCA: string,
    ociRepositories: string[],
    skipTLS: boolean,
    passCredentials: boolean,
  ) {
    const { data } = await axiosWithAuth.post<any>(
      url.backend.apprepositories.validate(cluster, namespace),
      {
        appRepository: {
          repoURL,
          type,
          authHeader,
          authRegCreds,
          customCA,
          ociRepositories,
          tlsInsecureSkipVerify: skipTLS,
          passCredentials: passCredentials,
        },
      },
    );
    return data;
  }
}
