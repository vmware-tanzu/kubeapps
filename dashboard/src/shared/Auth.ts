// Copyright 2018-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { grpc } from "@improbable-eng/grpc-web";
import { AxiosResponse } from "axios";
import * as jwt from "jsonwebtoken";
import { get } from "lodash";
import { IConfig } from "./Config";
import { KubeappsGrpcClient } from "./KubeappsGrpcClient";
import {
  InternalServerNetworkError,
  NotFoundNetworkError,
  UnauthorizedNetworkError,
} from "./types";
const AuthTokenKey = "kubeapps_auth_token";
const AuthTokenOIDCKey = "kubeapps_auth_token_oidc";

export class Auth {
  public static resourcesServiceClient = (token?: string) =>
    new KubeappsGrpcClient().getResourcesServiceClientImpl(token);

  public static getAuthToken() {
    return localStorage.getItem(AuthTokenKey);
  }

  public static setAuthToken(token: string, oidc: boolean) {
    localStorage.setItem(AuthTokenOIDCKey, oidc.toString());
    if (token) {
      localStorage.setItem(AuthTokenKey, token);
    }
  }

  public static unsetAuthToken() {
    localStorage.removeItem(AuthTokenKey);
  }

  public static unsetAuthCookie(config: IConfig) {
    // http cookies cannot be deleted (or modified or read) from client-side
    // JS, so force browser to load the sign-out URI (which expires the
    // session cookie).
    localStorage.removeItem(AuthTokenOIDCKey);
    window.location.assign(config.oauthLogoutURI || "/oauth2/sign_out");
  }

  public static usingOIDCToken() {
    return localStorage.getItem(AuthTokenOIDCKey) === "true";
  }

  public static wsProtocols() {
    const token = this.getAuthToken();
    // If we're using OIDC for auth, then let the auth proxy handle
    // injecting the ws creds.
    if (!token || this.usingOIDCToken()) {
      return [];
    }
    return [
      // Trimming the b64 padding character ("=") as it is not accepted by k8s
      // https://github.com/kubernetes/apiserver/blob/release-1.22/pkg/authentication/request/websocket/protocol.go#L38
      "base64url.bearer.authorization.k8s.io." +
        Buffer.from(token).toString("base64").replaceAll("=", ""),
      "binary.k8s.io",
    ];
  }

  public static fetchOptions(): RequestInit {
    const headers = new Headers();
    headers.append("Authorization", `Bearer ${this.getAuthToken()}`);
    return {
      headers,
    };
  }

  // Throws an error if the token is invalid
  public static async validateToken(cluster: string, token: string) {
    try {
      await this.resourcesServiceClient(token).CheckNamespaceExists({
        context: { cluster, namespace: "default" },
      });
    } catch (e: any) {
      if (e.code === grpc.Code.Unauthenticated) {
        throw new UnauthorizedNetworkError("invalid token");
      }
      // https://kubernetes.io/docs/reference/access-authn-authz/authentication/#anonymous-requests
      // Since we are always passing a token here, A 403 authorization error
      // only occurs if the token resulted in successful authentication. We
      // don't make any assumptions over RBAC for the requested namespace or
      // other required authz permissions until operations on those resources
      // are attempted (though we may want to revisit this in the future).
      if (e.code !== grpc.Code.PermissionDenied) {
        if (e.code === grpc.Code.NotFound) {
          throw new NotFoundNetworkError("not found");
        }
        if (e.code === grpc.Code.Internal) {
          throw new InternalServerNetworkError("internal error");
        }
        throw new InternalServerNetworkError(`${e.code}: ${e.message}`);
      }
    }
  }

  // isErrorFromAPIsServer returns true if the response is a 403 determined to have originated
  // from the grpc-web APIs server, rather than the auth proxy.
  public static isErrorFromAPIsServer(e: any): boolean {
    const contentType = e.metadata?.headersMap["content-type"] as string[];
    if (contentType.some(v => v.startsWith("application/grpc-web"))) {
      return true;
    }
    return false;
  }

  // is403FromAuthProxy returns true if the response is a 403 determined to have originated
  // from the auth proxy itself, rather than upstream.
  //
  // Ideally we would be able to set a header for responses generated by the
  // auth proxy, rather than rely on the fact that the 403 response sent by
  // the auth proxy is (by default) an html page (rather than the json
  // upstream result). Hence encapsulating this ugliness here so we can fix
  // it in the one spot. We may need to query `/oauth2/info` to avoid potential
  // false positives.
  // Note: This function is only used now by AxiosInstance which
  // in turn is only used for operators support.
  public static is403FromAuthProxy(r: AxiosResponse<any>): boolean {
    if (r.data && typeof r.data === "string" && r.data.match("system:serviceaccount")) {
      // If the error message is related to a service account is not from the auth proxy
      return false;
    }
    return r.status === 403 && (!r.data || !r.data.message);
  }

  // isAnonymous returns true if the message includes "system:anonymous"
  // in response.data or response.data.message
  // the k8s api server nowadays defaults to allowing anonymous
  // requests, so that rather than returning a 401, a 403 is returned if
  // RBAC does not allow the anonymous user access.
  //
  // Note: This function is only used now by AxiosInstance which
  // in turn is only used for operators support.
  public static isAnonymous(response: AxiosResponse<any>): boolean {
    const msg = get(response, "data.message") || get(response, "data");
    return typeof msg === "string" && msg.includes("system:anonymous");
  }

  // isAuthenticatedWithCookie() does a GET request to determine if
  // the request is authenticated with an http-only cookie (there is, by design,
  // no way to determine via client JS whether an http-only cookie is present).
  //
  // Note that when using the auth-proxy, anonymous requests never
  // get to the backend since the auth-proxy requires authentication.
  // But if this function is incorrectly called when the auth-proxy
  // is not in use, with a cluster that supports anonymous requests,
  // it could potentially return a false positive.
  public static async isAuthenticatedWithCookie(cluster: string): Promise<boolean> {
    try {
      await this.resourcesServiceClient().CheckNamespaceExists({
        context: { cluster, namespace: "default" },
      });
    } catch (e: any) {
      // The only error response which can possibly mean we did authenticate is
      // a 403 from the k8s api server (ie. we got through to k8s api server
      // but RBAC doesn't authorize us).
      if (e.code !== grpc.Code.PermissionDenied) {
        return false;
      }

      // A 403 error response from our APIs server, rather than the
      // auth proxy, means we are authenticated and did get
      // through to the API server but were rejected by RBAC.
      if (this.isErrorFromAPIsServer(e)) {
        return true;
      }
      return false;
    }
    return true;
  }

  // defaultNamespaceFromToken decodes a jwt token to return the k8s service
  // account namespace.
  // TODO(mnelson): until we call jwt.verify on the token during validateToken above
  // we use a default namespace for both invalid tokens and tokens without the expected
  // key.
  public static defaultNamespaceFromToken(token: string) {
    const payload = jwt.decode(token);
    const namespaceKey = "kubernetes.io/serviceaccount/namespace";
    if (payload && payload[namespaceKey]) {
      return payload[namespaceKey];
    }
    return "";
  }
}
