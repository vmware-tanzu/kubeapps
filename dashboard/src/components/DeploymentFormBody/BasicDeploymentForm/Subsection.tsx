// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { setValue } from "shared/schema";
import { DeploymentEvent, IBasicFormParam } from "shared/types";
import { getValueFromEvent } from "shared/utils";
import Param from "./Param";

export interface ISubsectionProps {
  label: string;
  param: IBasicFormParam;
  allParams: IBasicFormParam[];
  appValues: string;
  deploymentEvent: DeploymentEvent;
  handleValuesChange: (value: string) => void;
}

function Subsection({
  label,
  param,
  allParams,
  appValues,
  deploymentEvent,
  handleValuesChange,
}: ISubsectionProps) {
  const handleChildrenParamChange = (childrenParam: IBasicFormParam) => {
    return (e: React.FormEvent<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>) => {
      const value = getValueFromEvent(e);
      param.children = param.children!.map(p =>
        p.path === childrenParam.path ? { ...childrenParam, value } : p,
      );
      handleValuesChange(setValue(appValues, childrenParam.path, value));
    };
  };

  return (
    <div className="subsection">
      <div>
        <label className="deployment-form-label">{label}</label>
        {param.description && (
          <>
            <br />
            <span className="description">{param.description}</span>
          </>
        )}
      </div>
      {param.children &&
        param.children.map((childrenParam, i) => {
          const id = `${childrenParam.path}-${i}`;
          return (
            <Param
              param={childrenParam}
              allParams={allParams}
              id={id}
              key={id}
              handleBasicFormParamChange={handleChildrenParamChange}
              deploymentEvent={deploymentEvent}
              appValues={appValues}
              handleValuesChange={handleValuesChange}
            />
          );
        })}
    </div>
  );
}

export default Subsection;
