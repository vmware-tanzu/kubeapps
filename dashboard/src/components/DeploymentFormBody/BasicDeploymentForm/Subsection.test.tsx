// Copyright 2019-2022 the Kubeapps contributors.
// SPDX-License-Identifier: Apache-2.0

import { shallow } from "enzyme";
import { IBasicFormParam } from "shared/types";
import Subsection, { ISubsectionProps } from "./Subsection";

const defaultProps = {
  label: "Enable an external database",
  param: {
    children: [
      {
        path: "externalDatabase.database",
        type: "string",
        value: "bitnami_wordpress",
      },
      { path: "externalDatabase.host", type: "string", value: "localhost" },
      { path: "externalDatabase.password", type: "string" },
      { path: "externalDatabase.port", type: "integer", value: 3306 },
      {
        path: "externalDatabase.user",
        type: "string",
        value: "bn_wordpress",
      },
      {
        path: "mariadb.enabled",
        title: "Enable External Database",
        type: "boolean",
        value: true,
      } as IBasicFormParam,
    ],
    path: "externalDatabase",
    title: "External Database Details",
    description: "description of the param",
    type: "object",
  } as IBasicFormParam,
  allParams: [],
  appValues: "externalDatabase: {}",
  deploymentEvent: "install",
  handleValuesChange: jest.fn(),
} as ISubsectionProps;

it("should render a external database section", () => {
  const wrapper = shallow(<Subsection {...defaultProps} />);
  expect(wrapper).toMatchSnapshot();
});
