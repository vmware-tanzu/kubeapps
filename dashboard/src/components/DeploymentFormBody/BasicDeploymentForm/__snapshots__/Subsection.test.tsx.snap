// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a external database section 1`] = `
<div
  className="subsection"
>
  <div>
    <label
      className="deployment-form-label"
    >
      Enable an external database
    </label>
    <br />
    <span
      className="description"
    >
      description of the param
    </span>
  </div>
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="externalDatabase.database-0"
    key="externalDatabase.database-0"
    param={
      Object {
        "path": "externalDatabase.database",
        "type": "string",
        "value": "bitnami_wordpress",
      }
    }
  />
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="externalDatabase.host-1"
    key="externalDatabase.host-1"
    param={
      Object {
        "path": "externalDatabase.host",
        "type": "string",
        "value": "localhost",
      }
    }
  />
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="externalDatabase.password-2"
    key="externalDatabase.password-2"
    param={
      Object {
        "path": "externalDatabase.password",
        "type": "string",
      }
    }
  />
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="externalDatabase.port-3"
    key="externalDatabase.port-3"
    param={
      Object {
        "path": "externalDatabase.port",
        "type": "integer",
        "value": 3306,
      }
    }
  />
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="externalDatabase.user-4"
    key="externalDatabase.user-4"
    param={
      Object {
        "path": "externalDatabase.user",
        "type": "string",
        "value": "bn_wordpress",
      }
    }
  />
  <Param
    allParams={Array []}
    appValues="externalDatabase: {}"
    deploymentEvent="install"
    handleBasicFormParamChange={[Function]}
    handleValuesChange={[MockFunction]}
    id="mariadb.enabled-5"
    key="mariadb.enabled-5"
    param={
      Object {
        "path": "mariadb.enabled",
        "title": "Enable External Database",
        "type": "boolean",
        "value": true,
      }
    }
  />
</div>
`;
